### ISSUE TITLE
screen jumps to top before page change

### ISSUE THREAD

**Issue by maxim25** [nuxt/nuxt.js#2593]
# Issue
scrollToTop fires before page change

# To Reproduce
Create 2 pages, A and B, linking B to a with a nuxt-link. 
Make sure A is long enough to be able to scroll down.
When you scroll down on page A and click on the nuxt-link to go to page B, then page A will first scroll to the top before transitioning to page B. 

I am using Chrome Version 63.0.3239.132

My issue seems to be the same as #1036 but it was closed?? Any update on this?

**Comment by maxim25** [comment]
I added the following to my nuxt.config.js. It doesn't fix the back button jump, but still gives a better user experience

```router: {
         scrollBehavior: function (to, from, savedPosition) {
             return {x: false, y: false}
         }
     },
```

**Comment by colloquet** [comment]
Hi, I am also having the same problem, I noticed that the scroll function is called before the page is mounted. Whereas it's the other way round for my other non-ssr vue apps.

**Comment by bjunc** [comment]
I think the desired behavior would be that the `scrollToTop` would be triggered _after_ page A transitions out, and _before_ page B transitions in:

Current sequence:

1. User scrolls down PageA and clicks a `<nuxt-link/>` for PageB
1. Nuxt fires `scrollToTop`; which shows the top of PageA
1. PageA transitions out (eg. fade out)
1. PageB transitions in (eg. fade in)

Ideal sequence:

1. User scrolls down PageA and clicks a `<nuxt-link/>` for PageB
1. PageA transitions out (eg. fade out)
1. Nuxt fires `scrollToTop`.  No "page" is shown because we're between transitions
1. PageB transitions in (eg. fade in)

I think a lot of this has to do with the Vue Router flow (and maybe less to do with Nuxt).  Below is the navigation flow.  Not sure where the scroll behavior / transitions fit into this.

1. The Full Navigation Resolution Flow
1. Navigation triggered.
1. Call leave guards in deactivated components.
1. Call global beforeEach guards.
1. Call beforeRouteUpdate guards in reused components (2.2+).
1. Call beforeEnter in route configs.
1. Resolve async route components.
1. Call beforeRouteEnter in activated components.
1. Call global beforeResolve guards (2.5+).
1. Navigation confirmed.
1. Call global afterEach hooks.
1. DOM updates triggered.
1. Call callbacks passed to next in beforeRouteEnter guards with instantiated instances.

**Comment by husayt** [comment]
This should be resolved with latest nuxt. If not please update us.

**Comment by maxim25** [comment]
@husayt just checked it out, and yes the screen jump is gone! However now it doesn't go to the top of the 'to' page.

**Comment by onlyurei** [comment]
A working work-around fix of using a custom `scrollBehavior` in `route` option in `nuxt.config.js`: https://github.com/nuxt/nuxt.js/issues/2738#issuecomment-372007743.

**Comment by studioscholz** [comment]
@husayt After updating to 1.4 the screen jump is gone in dev mode, however it still persists in a production built prototype: http://www.soccochico.com/

**Comment by raphaelaitelalim** [comment]
None of the solutions proposed above have worked for me, but wrapped my container in <no-ssr/> component worked ! Hope this will help someone !

**Comment by RandomPiche** [comment]
@raphaelaitelalim What do you mean by "wrapped my container in component" ? What container / component are you talking about ? We don't even know the structure you're working on.

**Comment by raphaelaitelalim** [comment]
@RandomPiche Sorry about that, so i wanted to say "wrapped my container in <no-ssr/> component".

My code look like this : 

`<template>
     <no-ssr>
        <div>
            some content...
        </div>
      </no-ssr>
</template>`

**Comment by saicharan96** [comment]
I have same issue @raphaelnoguier still after wrapped my container in component

**Comment by raphaelnoguier** [comment]
@saicharan96 Okay can i see your layout/default.vue and nuxt config ? 

Thanks

**Comment by cdefy** [comment]
Hello there,

I know this issue is marked as closed but I just had the same problem. I'm up to date with Nuxt version, I tried the different solutions proposed here without success.

Worst is I have the bug on a remote environment but not when I run the web-app locally. It appears once then disappears without me doing anything. I'm a bit lost :o 

Thanks for your help!

**Comment by JoaquimLey** [comment]
Indeed, this is **still an issue** even with the latest version.
To note I'm using the universal/SSR mode.

**Comment by Razz21** [comment]
My workaround for this problem (in some cases) is page transition with 'in-out' mode, set `scrollToTop` to `false` in component, and manually update `document.documentElement.scrollTop` after trasition ends.

```vue
<script>
//page component
export default {
  scrollToTop: false, //disable this option in component
 
  transition: {
    name: "page",
    mode: "in-out",
    afterEnter(el) {
      // reset scroll position
      document.documentElement.scrollTop = 0;
      ...
    }
  }
}
</script>
<style>
/* 
=================
example page transition
=================
*/

.page-enter-active,
.page-leave-active {
  overflow: hidden;
  position: fixed;
  top: 0;
  left: 0;
  width: 100vw;
  height: 100vh;
  visibility: hidden; 
  z-index: 200;
}
</style>
```

This is not very flexible soluion, not tested with async data, but maybe help someone. [Example](https://codesandbox.io/s/shy-violet-uck4k)

**Comment by pooledge** [comment]
My workaround:
- In your layout, surround `<Nuxt />` component with a `<transition />`
- Attach a function emitting e.g. 'page-after-leave' to `$root` on `@after-leave` hook
- In `/app/router.scrollBehavior.js` listen to the emit:

```
export default (to, from, savedPosition) => {
	const nuxt = window['<%= globals.nuxt %>']; // in case you don't use the default $nuxt

	let position = { // default, mimics the top position on navigation
		x: 0,
		y: 0,
	};

	return new Promise((resolve) => {
		nuxt.$root.$once('page-after-leave', () => {
			nuxt.$nextTick(() => {
				resolve(position);
			});
		});
	});
};
```

It's a bit on overhead with additional `<transition />` but it works very stable, no jumps!

**Comment by tre-dev** [comment]
> Attach a function emitting e.g. 'page-after-leave' to $root on @after-leave hook

@pooledge Can you please give an example?

**Comment by Razz21** [comment]
@tre-dev I am also using this solution and works fine so far. Just emit custom event in `@after-leave` transition hook, and listen for it in custom `/app/router.scrollBehavior.js` file as described above. `$root` is global object and can be used as global event bus.

```vue

<script>
export default {
  ... 
  transition: {
    name: "page",
    ...
    afterLeave(el) {
      this.$root.$emit('page-after-leave');
      ...
    }
    ...
  }
}
</script>
```

**Comment by EdisonHarada** [comment]
O'm not sure if there's someone else looking into it but the way that I used to fix is by adding this scroll behavior to my router:

```
scrollBehavior(to, from, savedPosition) {
      const nuxt = window['<%= globals.nuxt %>']

      if (to.path === from.path) {
        return
      }
      if (to.hash) {
        return new Promise((resolve, reject) => {
          setTimeout(() => {
            resolve({
              selector: to.hash
            })
          }, 500)
        })
      }

      const position = savedPosition || { x: 0, y: 0 }
      return new Promise((resolve, reject) => {
        nuxt.$root.$once('triggerScroll', () => {
          nuxt.$nextTick(() => { resolve(position) })
        })
      })
    }
  })
```

Instead of using "page-after-leave" as @pooledge used, I updated to use "triggerScroll" from nuxt, because nuxt is triggering this event even when we change the layout.

**Comment by whoisjorge** [comment]
It's probably unrelated, but check if you don't have any `smooth-scroll: smooth;` attached to an element.

**Comment by salzig** [comment]
removing 
```
body {
  overflow-x: hidden;
  overflow-y: scroll;
}
```
solved the problem for us.

**Comment by vinuv00702** [comment]
iam having the same issue? have anyone got a fix yet?

**Comment by desanvicenteg** [comment]
El AltepeMerario GDSV 
![Cargando 147891615...]

**Comment by tomaslucena** [comment]
did someone ever fix this? I installed nuxt 2 weeks ago and the problem is still there

**Comment by tubstrr** [comment]
I found a solve for my issue.

On my site, I have some pretty complicated screen transitions, and a few DOM elements that on `mount()` get sizing information from `$refs` and the distance to the top of the window. I had to have the entering page be `position: fixed; top:0;` in order for my calculations to work.

I realized that I also set the leaving page to `position: fixed;`, which was the troublesome part for me.

Setting `/app/router.scrollBehaviour.js` to:

` export default function (to, from, savedPosition) {
  return false;
}
`

achived the behavior that I wanted.



To summarize, I had 2 issues.  1) Router behavior was set to the default Scroll To Top on route change.  2) My leaving page was set to `position: fixed; top: 0;`
